namesys/namesys.go

```diff
@@ -72,37 +94,85 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
 		return "", ErrResolveFailed
 	}
 
-	for protocol, resolver := range ns.resolvers {
-		log.Debugf("Attempting to resolve %s with %s", segments[2], protocol)
-		p, err := resolver.resolveOnce(ctx, segments[2])
-		if err == nil {
-			if len(segments) > 3 {
-				return path.FromSegments("", strings.TrimRight(p.String(), "/"), segments[3])
-			} else {
-				return p, err
+	makePath := func(p path.Path) (path.Path, error) {
+		if len(segments) > 3 {
+			return path.FromSegments("", strings.TrimRight(p.String(), "/"), segments[3])
+		} else {
+			return p, nil
+		}
+	}
+
+	// Resolver selection:
+	// 1. if it is a multihash resolve through "pubsub" (if available),
+	//    with fallback to "dht"
+	// 2. if it is a domain name, resolve through "dns"
+	// 3. otherwise resolve through the "proquint" resolver
+	key := segments[2]
+
+	_, err := mh.FromB58String(key)
+	if err == nil {
+		res, ok := ns.resolvers["pubsub"]
+		if ok {
+			p, err := res.resolveOnce(ctx, key)
+			if err == nil {
+				return makePath(p)
+			}
+		}
+
+		res, ok = ns.resolvers["dht"]
+		if ok {
+			p, err := res.resolveOnce(ctx, key)
+			if err == nil {
+				return makePath(p)
+			}
+		}
+
+		return "", ErrResolveFailed
+	}
+
+	if isd.IsDomain(key) {
+		res, ok := ns.resolvers["dns"]
+		if ok {
+			p, err := res.resolveOnce(ctx, key)
+			if err == nil {
+				return makePath(p)
 			}
 		}
+
+		return "", ErrResolveFailed
 	}
+
+	res, ok := ns.resolvers["proquint"]
+	if ok {
+		p, err := res.resolveOnce(ctx, key)
+		if err == nil {
+			return makePath(p)
+		}
+
+		return "", ErrResolveFailed
+	}
+
 	log.Warningf("No resolver found for %s", name)
 	return "", ErrResolveFailed
 }
 
 // Publish implements Publisher
 func (ns *mpns) Publish(ctx context.Context, name ci.PrivKey, value path.Path) error {
-	err := ns.publishers["/ipns/"].Publish(ctx, name, value)
-	if err != nil {
-		return err
-	}
-	ns.addToDHTCache(name, value, time.Now().Add(DefaultRecordTTL))
-	return nil
+	return ns.PublishWithEOL(ctx, name, value, time.Now().Add(DefaultRecordTTL))
 }
 
 func (ns *mpns) PublishWithEOL(ctx context.Context, name ci.PrivKey, value path.Path, eol time.Time) error {
-	err := ns.publishers["/ipns/"].PublishWithEOL(ctx, name, value, eol)
+	err := ns.publishers["dht"].PublishWithEOL(ctx, name, value, eol)
 	if err != nil {
 		return err
 	}
 	ns.addToDHTCache(name, value, eol)
+
+	pub, ok := ns.publishers["pubsub"]
+	if ok {
+		pub.PublishWithEOL(ctx, name, value, eol)
+	}
```

Comment:
Thinking some more about this, the pubsub Publish can be blocking as it might have to perform bootstrap the first time it publishes something; this can take a few seconds.

So yeah, I will parallelize the publish. It's worth the complexity, since both can be potentially slow.